doc/user-manual/language/opaque-definitions.lagda.rst

```diff
@@ -98,38 +98,140 @@ module, say ``Integer-ring``, can provide its own proofs, in an
-with an ``opaque`` block, each opaque name that ``X`` (transitively)
-unfolds will also be marked transparent. This is necessary to preserve
-subject reduction. Consider the following example::
+When an ``opaque`` block is checked, Agda will compute ahead-of-time the
+set of names it is allowed to unfold. This set is per-*block*, not
+per-*definition*. An ``unfolding`` clause, if it mentions opaque names,
+will cause the unfolding sets associated with those names to be added to
+the current block.
 
-  opaque
-    foo : Set
-    foo = Nat
+The following illustrates the behaviour of these rules:
 
-  opaque
-    unfolding foo
+- Unfolding any name in an opaque block will cause any of the *other*
+  names in that block to be unfolded as well. Example::
 
-    bar : foo
-    bar = 123
+    module _ where private
+      opaque
+        x : Nat
+        y : Nat
 
-If we could unfold ``bar`` independently of ``foo``, we'd end up with
-``bar : foo`` reducing to ``123 : Nat``, while in this block, ``foo !=
-Nat``. One can observe the transitivity of ``unfolding`` clauses with an
-example like this::
+        x = 3
+        y = 4
 
-  opaque
-    unfolding bar
+      opaque
+        unfolding x
+
+        _ : y ≡ 4
+        _ = refl
+
+  Here, even though only ``x`` was asked for, ``y`` is also available
+  for unfolding.
+
+- Since the unfolding sets brought in by clauses are associated with the
+  block, unfolding is transitive::
+
+    module _ where private
+      opaque
+        x : Nat
+        x = 3
+
+      opaque
+        unfolding x
+        y : Nat
+        y = 4 + x
+
+      opaque
+        unfolding y
+        _ : y ≡ 7
+        _ = refl
+
+- Opaque blocks which are lexically nested can also unfold the names of
+  their *parent* blocks, even if the name is not in scope when the child
+  block is defined::
+
+    module _ where private
+      opaque
+        x : Nat
+        x = 3
+
+        opaque
+          y : Nat
+          y = 4
+
+          _ : x ≡ 3
+          _ = refl
+
+        z : Nat
+        z = 5
+
+      opaque
+        unfolding y
+        _ : z ≡ 5
+        _ = refl
+
+  This is because the ``x`` and ``z`` are direct children of the same
+  ``opaque`` block: the ``opaque`` block that defines ``y`` does not
+  "split" its parent block.
+
+Multiple unfolding clauses are supported, as well as unfolding more than
+one name per clause. The syntax for the latter is simply a
+space-separated list of names, which must refer to unambiguous
+functions::
+
+    module _ where private
+      opaque
+        x : Nat
+        x = 3
+
+      opaque
+        y : Nat
+        y = 4
+
+      opaque
+        z : Nat
+        z = 5
+
+      opaque
+        unfolding x y
+        unfolding z
+
+        _ : x + y + z ≡ 12
+        _ = refl
+
+Note finally that the names in ``unfolding`` clauses are not affected by
+layout, so the following is legal. This allows the user to choose their
```

Comment:
The following is not allowed:
```agda
opaque
    unfolding A
  B
```
Thus the names are, in some sense, affected by layout.